Updates to FUSE surface runoff parameterizations (PRMS and ARNO/VIC) and surface runoff combinations

seantrim opened the pull request:
Hi @ashleymedin - Thanks for your recent updates. It took a while to combine things but I think the code should be significantly improved.

This PR introduces updates to the FUSE PRMS and ARNO/VIC surface runoff parameterizations as requested by Martyn and Wouter, with adaptations for your recent changes in `soilLiqFlx` operations. Additionally, combinations of surface runoff methods are now possible.

The model decision structure for the liquid flux boundary condition structure has been changed to permit combinations of methods. If `bcUpprSoiH` = `liq_flux` , the `surfRun_IE` and `surfRun_SE` model decisions specify which runoff methods to use for infiltration excess and saturation excess runoff methods, respectively. The options for `surfRun_IE` include `zero_IE` and `homegrown_IE` (default) . Options for `surfRun_SE` include `zero_SE` , `homegrown_SE` (default) , `FUSEPRMS` , `FUSEAVIC` , and `FUSETOPM` (updates for the TOPMODEL option will be sent in an additional PR). The homegrown options are for SUMMA's original liquid flux method (divided into infiltration and saturation excess components), which are the default values to preserve functionality of existing workflows. The zero options specify zero surface runoff for the selected runoff component.

The infiltration and saturation excess surface runoff components are gathered near the end of `surfaceFlx` in `update_gather_runoff_components` . This includes evaluating the total surface runoff, total surface infiltration, and total surface infiltration derivatives. The `scalarSoilControl` value is also evaluated there based on the total infiltration. **Please check `scalarSoilControl` to make sure it is what you wanted.** The current method for `scalarSoilControl`  works for all test suite cases but I was not completely certain of it.

The updates to FUSE PRMS and ARNO/VIC include the application of the `LogSumExp` function for smooth approximations to minimum functions. This impacted the infiltration derivatives as well (`SoftArgMax` is the derivative of `LogSumExp` which now appears in the derivative formulas). The `LogSumExp` and `SoftArgMax` functions rely on a smoothing parameter `alpha_LSE`. For now, that parameter is set to 5x10^5, which resulted in more than enough accuracy in my tests. However, there may be accuracy issues for extremely dry soil layers. I put in a check with an error message if this problem ever occurs.

I have also applied the PRMS and ARNO/VIC parameterizations to individual soil layers rather than the integrated soil column. This was done to prevent very large derivative magnitudes occurring. In past comments, Martyn has indicated that this localized approach is also reasonable. Updates to TOPMODEL are pending.

This branch was successfully tested using the test suite plus additional cases that use the FUSE parameterizations and runoff method combinations. Test suite output varied by a minor amount compared to previous solutions. Part of the difference is due to changing how the order of operations for computing total runoff/infiltration values. Part of the difference is also your recent derivative updates (and increased evaluation frequency).

Make sure all the relevant boxes are checked (and only check the box if you actually completed the step):

- [ ] Closes #xxx (identify the issue associated with this PR)
- [x] Code passes standard test cases (results are either bit-for-bit identical, or differences are explained in the PR comment)
- [ ] New tests added (describe which tests were performed to test the changes)
- [ ] Science test figures (add figures to PR comment and describe the tests)
- [ ] Checked that the new code conforms to the [SUMMA coding conventions](https://github.com/NCAR/summa/blob/master/docs/howto/summa_coding_conventions.md)
- [ ] Describe the change in the release notes (use  `./summa/docs/whats-new.md`)

wknoben committed:
Merge pull request #590 from ashleymedin/develop

sundials develop continuing merge

wknoben committed:
Merge pull request #595 from ashleymedin/develop

Continuing Sundials merge

wknoben committed:
Merge pull request #597 from ashleymedin/develop

patch for excess infiltration

wknoben committed:
fixed version and git info not being collected properly during compile

wknoben committed:
Merge pull request #599 from wknoben/bugfix/cmake_version_info

fixed version and git info not being collected properly during compile

wknoben committed:
clarified model decision error messages with exact wording needed in mDecisions file

wknoben committed:
Merge pull request #604 from wknoben/decision-error-clarification

clarified model decision error messages with exact wording needed in mDecisions file

wknoben committed:
bugfix to have Smax calculations account for porosity of soil column

wknoben committed:
Merge pull request #606 from wknoben/fuse_depth_fix

Bugfix: Smax calculations in FUSE ARNO/VIC and PRMS saturation excess options

seantrim committed:
Added comments to clarify usage of iLayerHeight array in soilLiqFlx.

seantrim committed:
Implemented model decisions and corresponding options for infiltration and saturation excess surface runoff methods.

seantrim committed:
Added infiltration and saturation excess surface runoff model decisions to surfaceFlx input object class.

seantrim committed:
Implementation of infiltration and saturation excess surface runoff logic in surfaceFlx in progress.

seantrim committed:
Implemention of individual surface runoff component variables in progress -- zero_IE, zero_SE, and FUSEPRMS completed.

seantrim committed:
Implemented infiltration and saturation excess component variables for FUSE ARNO/VIC and FUSE TOPMODEL surface runoff parameterizations.

seantrim committed:
Updates to infiltration derivative components for infiltration and saturations excess surface runoff components in surfaceFlx.

seantrim committed:
Updates to infiltration derivative calculations and comments in surfaceFlx.

seantrim committed:
Made homegrown surface runoff procedure the default for infiltration ans saturation runoff methods for all solver choices.

seantrim committed:
Merge branch 'develop' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

seantrim committed:
Implemented LogSumExp function for smoothed max/min function evaluations in soil_utils_module.

seantrim committed:
Implemented smoothed minimum for FUSE PRMS -- infiltration derivatives in progress.

seantrim committed:
Merge branch 'develop' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

seantrim committed:
Added SoftArgMax function to soil_utils_module.

seantrim committed:
Updates to derivative calculations for FUSE PRMS.

ashleymedin committed:
typos

ashleymedin committed:
remove deriv_desired since it's always .true.

ashleymedin committed:
fixing derivs

seantrim committed:
Merge branch 'develop' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

seantrim committed:
Merge branch 'fix_FUSEderiv2' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

seantrim committed:
Merge branch 'develop' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

ashleymedin committed:
groundwater doesn't use matricHeadLiq

ashleymedin committed:
put derivatives in with computInfil flag

ashleymedin committed:
fix, dq_dNrgStateUnsat derivative that was missing, not sure it's right though

ashleymedin committed:
lost bit

seantrim committed:
Added quadruple-precision operations to LogSumExp and SoftArgMax to permit a great range of smoothing parameters.

seantrim committed:
Updates to FUSE PRMS and FUSE ARNO/VIC: parameterizations now applied locally to each soil layer, smoother functions were applied, and derivatives were updated.

ashleymedin committed:
fixing the infiltration derivatives and the dense snow derivatives

ashleymedin committed:
remove a comment

ashleymedin committed:
Merge branch 'develop' into fix_FUSEderiv2

* develop:
  variable add in summa_init
  remove unneeded line
  removing unneeded type
  minExpLogHgt can't be a global data because it's not constant
  more typos in comments
  remove greenVegFrac_monthly as a variable (repeat) and move minExpLogHgt global data calculation into summa_setup with all the others
  removing unused stuff

ashleymedin committed:
removed a print

seantrim committed:
Merge branch 'develop' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

ashleymedin committed:
a bit more on Jacobian

seantrim committed:
Merge branch 'fix_FUSEderiv2' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

seantrim committed:
Merge branch 'fix_FUSEderiv2' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

seantrim committed:
Updated computation of scalarSurfaceControl to compensate for combinations of runoff methods.

seantrim committed:
Adjusted method for checking the sum of infiltration and saturation excess runoff components.

seantrim committed:
Updates for accuracy of LogSumExp and SoftArgMax function evaluations.

ashleymedin committed:
remove a print

ashleymedin commented:
The scalarSoilControl looks correct to me. I think a lot of the comments could be deleted for conciseness. It seems fine to just define scalarSoilControl instead of commenting on it's value in the subroutines.

ashleymedin commented:
"applied the PRMS and ARNO/VIC parameterizations to individual soil layers rather than the integrated soil column. This was done to prevent very large derivative magnitudes occurring. In past comments, Martyn has indicated that this localized approach is also reasonable", -- but liq_flx is applied to the whole rootZone in the soil column so this seems odd; seems they should be the same.

ashleymedin commented:
I don't think these derivatives are correct. If you only look at the first layer, then the derivative should only be on the first layer. I also don't see where the derivatives are blowing up to warrant an expensive smoothing operation and/or just using the top layer, and when I tested them with the analytical solution they are quite small in every layer. Could we try using the actual values? dVolFracLiq_dWat and dVolFracLiq_dTk are used all over the code without smoothing. 

I can see a case for S1_T=min(S1,S1_T_max), Ac = (S1_T/S1_T_max)*Ac_max needing a smoother (FUSE PRMS), but Ac = 1._rkind - (1._rkind-S1/S1_max)**b (FUSE ARNO VIC) will never have an negative value going **b since S1/S1_max<=1.

Also, it seems the derivatives are all missing a factor of mLayerDepth(:).??

ashleymedin commented:
```
   dq_dHydStateVec(1:nSoil) = dq_dHydStateVec_IE(1:nSoil) + dq_dHydStateVec_SE(1:nSoil) ! infiltration derivative w.r.t hydrology state variable 
   dq_dNrgStateVec(1:nSoil) = dq_dNrgStateVec_IE(1:nSoil) + dq_dNrgStateVec_SE(1:nSoil) ! infiltration derivative w.r.t energy state variable 
```
Can be 
```
   dq_dHydStateVec(:) = dq_dHydStateVec_IE(:) + dq_dHydStateVec_SE(:) ! infiltration derivative w.r.t hydrology state variable 
   dq_dNrgStateVec(:) = dq_dNrgStateVec_IE(:) + dq_dNrgStateVec_SE(:) ! infiltration derivative w.r.t energy state variable 
```
along with a few others, 1:nSoil shouldn't be necessary in the code. 
Also, is it necessary to separate dq_dNrgStateVec_IE and dq_dNrgStateVec_SE if we don't save them and we will just combine them at the end of the code?

seantrim committed:
Merge branch 'fix_FUSEderiv2' of https://github.com/ashleymedin/summa into develop_sundials_FUSE

seantrim committed:
Removed extraneous comments on scalarSoilControl.

seantrim committed:
Updated FUSE PRMS to use uppermost layer for derivatives.

seantrim committed:
Updated FUSE ARNO/VIC to use uppermost soil layer for flux derivatives and made smoothing optional (currently off).

seantrim committed:
Simplified array slice notation in surfaceFlx.

seantrim commented:
I think I've addressed the comments above with these new commits. Usage of a smoothed minimum in ARNO/VIC is now controlled by a logical flag in case we wish to use it (currently set to off). The smoothed minimum in PRMS was retained. Both PRMS and ARNO/VIC derivatives were adjusted to use the top layer only.

Also, it seems that we need the dq_dNrgStateVec_IE and dq_dNrgStateVec_SE arrays for temporary storage, so those were retained.

ashleymedin commented:
Looks pretty good --  a few more things. I don't see dq_dHydStateVec_SE and dq_dNrgStateVec_SE getting initialized to 0 and we only set the (1) entry in the FUSE models; I might be missing these but seems they need to be set to 0 first. Second, I still don't see the reason to calculate dq_dHydStateVec_IE and dq_dHydStateVec_SE or the Nrg ones, since we don't take them outside the subroutine.

seantrim commented:
For the infiltration derivative components `dq_dHydStateVec_IE`, `dq_dHydStateVec_SE`, `dq_dNrgStateVec_IE`, and `dq_dNrgStateVec_SE`:

They are initialized to zero in `initialize_surfaceFlx`. So only assigning values for the uppermost soil layer in PRMS and ARNO/VIC should work.

The arrays are only in scope within `surfaceFlx`.

It may be possible to evaluate  `dq_dHydStateVec` and `dq_dNrgStateVec` without the IE and SE component arrays if we evaluate in place and are really careful about the order of operations. For 100 soil layers, this would save 2 * 100 * 8 bytes / 1024 bytes/KB = 1.5626 KB of memory per HRU (within `surfaceFlx`). We would still need the same number of floating point operations because we would still need the additions of IE and SE components (but stored in place).

However, I think this adds complexity to the algorithm that only you and I would know about (before we forget lol) for insignificant savings. For clarity and extensibility, I vote for retaining the IE and SE component arrays. The extra clarity on derivative evaluations reduces the chance of mistakes and may save development time (speaking from recent experience).

ashleymedin commented:
If I'm reading this right, dq_dNrgStateVec_IE  and dq_dHydStateVec_IE are only possibly non-zero if liq_flux and we get them from dq_dNrgStateVec and dq_dHydStateVec by subtracting a calculated dq_dNrgStateVec_SE  and dq_dHydStateVec_SE, and then throw away *_IE and *_SE. So to me it seems like an extraneous calculation to have them around. Is this behavior needed for future development? I agree that the expense savings is not much, but it's adding code complexity and messiness to have all these non-needed variables defined. 

Similarly, defining p = scalarRainPlusMelt, scalarSurfaceRunoff = qsx , and  scalarSurfaceInfiltration = infiltration makes the code longer, harder to read/decipher, and harder to debug, versus just using the names of the variables straight-up. Is this necessary?

seantrim committed:
Clean up of FUSE PRMS and ARNO/VIC to remove extraneous variables.

seantrim committed:
Simplify associate statements for FUSE parameterizations.

seantrim committed:
Added FUSE paper doi for reference for PRMS and ARNO/VIC.

seantrim committed:
Updated FUSE TOPMODEL surface runoff to use second soil layer as the lower FUSE layer.

seantrim commented:
Sorry about the overlapping variables. The extra notation came from the original FUSE paper so I could keep things straight while implementing the formulas. I've removed overlapping variables in favour of the original SUMMA variables (e.g., `scalarRainPlusMelt`) where possible. I've also cleaned up the associate blocks and comments.

So the extra business of computing the dq*_SE and dq*_IE variables in SUMMA's original liquid flux (which is now homegrown in this branch) was so that it is possible to use one of the runoff components in combination with another runoff method. For example, using homegrown for infiltration excess (IE) runoff and FUSE PRMS for saturation excess (SE) runoff. I know the separation adds some steps that weren't in the original liquid flux, but that was the reasoning behind it. So if you use homgrown for IE and SE, the original liquid flux results are obtained with some extra steps. However, overall run time did not change in my tests (and output matches within machine epsilon). I can see how just looking at liquid flux is now more confusing. My apologies for that.

The  dq*_SE and dq*_IE variables are non-zero for the first layer in PRMS and ARNO/VIC (and in the second layer in TOPMODEL).

I can't see an easy way to eliminate the dq*_SE and dq*_IE variables without messing with the functionality of combining different runoff methods for infiltration and saturation excess. However, I trust you - if you are seeing something I'm missing please feel free to make changes.

Oh yeah, I've added updates for the FUSE TOPMODEL parameterization. I've tried to follow the same steps we discussed for PRMS and ARNO/VIC. The FUSE layer 2 (`S2`) is now defined using soil layer 2 instead of the aquifer. I did not apply any smoothing for now. I know the calculations are complicated looking. The challenge is that the saturated area depends on the Gamma distribution, which makes some of the computations nasty. It's probably tough to follow what is going on from the code alone, but I did include a doi for the paper so people can look at things in detail if they want.  I feel like the original liquid flux method could also use a paper reference or something so people could see where the code is coming from.

I ran the test suite again (the official test suite plus some custom FUSE and runoff combination ones). The only change in output compared to my last reference solutions was for the TOPMODEL test case. The memory use and run times did not change for the official test suite compared to your develop branch.

ashleymedin commented:
Thanks for the details explanation and changes. I merged this. As for the gamma function ... yes a nasty derivative. We currently have a comment in the liq_flux section 
```
     !if we use this, we will have a derivative of scalarFrozenArea w.r.t. water and temperature in each layer (through mLayerVolFracIce)
     ! Should fix to deal with frozen area in the root zone, calculations would be expensive
     !scalarFrozenArea = 1._rkind - gammp(alpha,xLimg)      ! fraction of frozen area
```
so that had the same problem hence the scalarFrozenArea always being 0. I think it would be worth trying this also with the gamma function and the gamma derivative (and perhaps also using your smoother) on this section. People often comment on the fact that the frozen area is just 0 so I think it would be appreciated. (Unless we are planning to delete liq_flux altogether.)
